drake/systems/plants/RigidBodyManipulator.h

```diff
@@ -97,46 +99,256 @@ class DLLEXPORT_RBM RigidBodyManipulator
+                auto dSdotVdvi = dSdotv.rightCols(joint.getNumVelocities());
+                djoint_acceldv.rightCols(nv_joint) += transformSpatialMotion(element.transform_to_world, dSdotVdvi);
+
+                auto dJdotVdv = element.motion_subspace_in_world_dot_times_v.gradient().value().rightCols(num_velocities);
+                auto dJdotVdv_parent = parent_element.motion_subspace_in_world_dot_times_v.gradient().value().rightCols(num_velocities);
+                dJdotVdv.setZero();
+                for (int j = 0; j < nv_branch; j++) {
+                  int v_index = v_indices[j];
+                  dJdotVdv.col(v_index) = dJdotVdv_parent.col(v_index) + djoint_acceldv.col(j);
+                }
+              }
+            }
+          }
+        }
+      }
+      else {
+        element.transform_to_world.setIdentity();
+        // motion subspace in body frame is empty
+        // motion subspace in world frame is empty
+        // qdot to v is empty
+        // v to qdot is empty
+        if (compute_gradients) {
+          // gradient of transform
+          element.dtransform_to_world_dq.setZero();
+          // gradient of motion subspace in world is empty
+        }
+        if (cache.hasV()) {
+          element.twist_in_world.value().setZero();
+          element.motion_subspace_in_body.value().setZero();
+          element.motion_subspace_in_world.value().setZero();
+          element.qdot_to_v.value().setZero();
+          element.v_to_qdot.value().setZero();
+
+          if (compute_gradients) {
+            element.twist_in_world.gradient().value().setZero();
+            element.motion_subspace_in_body.gradient().value().setZero();
+            element.motion_subspace_in_world.gradient().value().setZero();
+            element.qdot_to_v.gradient().value().setZero();
+            element.v_to_qdot.gradient().value().setZero();
+          }
+          if (compute_JdotV) {
+            element.motion_subspace_in_body_dot_times_v.value().setZero();
+            element.motion_subspace_in_world_dot_times_v.value().setZero();
+            if (compute_gradients) {
+              element.motion_subspace_in_body_dot_times_v.gradient().value().setZero();
+              element.motion_subspace_in_world_dot_times_v.gradient().value().setZero();
+            }
+          }
+        }
+      }
+    }
+
+    cache.setJdotVCached(compute_JdotV && cache.hasV());
+  };
+
+  bool isBodyPartOfRobot(const RigidBody& body, const std::set<int>& robotnum) const;
+
+  double getMass(const std::set<int>& robotnum = RigidBody::defaultRobotNumSet) const;
 
   template <typename Scalar>
-  GradientVar<Scalar, SPACE_DIMENSION, 1> centerOfMass(int gradient_order, const std::set<int>& robotnum = RigidBody::defaultRobotNumSet);
+  GradientVar<Scalar, SPACE_DIMENSION, 1> centerOfMass(KinematicsCache<Scalar>& cache, int gradient_order, const std::set<int>& robotnum = RigidBody::defaultRobotNumSet) const;
```

Comment:
> I don't really like either name actually. The Kinematics part isn't really accurate. As far as Cache vs. Solution, I like Cache slightly better, but not super opinionated either way.
> 
I'm also ok, but would vote for Solution.  I don’t love it, but don’t feel like Cache is right.
> As for moving all of the methods from RBM to Kinematics*, that kind of makes sense. @patmarion <https://github.com/patmarion> mentioned that too. I'd like to do it in a future pull request though (and might abondon it if I run into unforseen trouble).
> 
Sounds perfect